subsys/net/lib/dns/dns_sd.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef DNS_SD_H_
+#define DNS_SD_H_
+
+#include <net/dns_sd.h>
+
+#include <stdint.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Iterate over read-only DNS-SD records
+ *
+ * The ability to iterate over DNS-SD records should
+ * only ever be used by system services.
+ *
+ * Usage:
+ * @code{c}
+ * const struct _dns_sd_rec *it;
+ * DNS_SD_FOREACH(it) {
+ *   do_something_with(it);
+ * }
+ * @endcode{c}
+ */
+#define DNS_SD_FOREACH(it) \
+	extern const struct _dns_sd_rec __start_dns_sd_rec; \
+	extern const struct _dns_sd_rec __stop_dns_sd_rec; \
+    for (it = &__start_dns_sd_rec; it < &__stop_dns_sd_rec; ++it)
+
+/**
+ * Handle a DNS Service Discovery PTR query
+ *
+ * On success, this function returns a positive, non-zero integer
+ * representing the size of the data in the response buffer, @p buf.
+ *
+ * When a DNS-SD enabled system receives a PTR query, it must
+ * examine the labels contained within the query and respond with
+ * appropriate TXT, SRV, and A / AAAA records if a match is found.
+ *
+ * This function should be called once for each DNS-SD record
+ * that matched a particular query. For each iteration, a separate
+ * response should be sent.
+ *
+ * For details, @see <a http="">RFC 6763</a>, Section 12.1.
+ *
+ * @param inst the @ref _dns_sd_sec representing the service
+ * @param ipv6 true when @p addr points to an IPv6 address
+ * @param addr a pointer to an IPv6 (uint8_t[16])or IPv4 (uint32_t) address
+ * @param buf the response buffer
+ * @param buf_size size of the response buffer
+ * @return a positive, non-zero value on success
+ * @return a negative errno value on failure
+ */
+int handle_ptr_query(const struct _dns_sd_rec *inst, bool ipv6,
```

Comment:
This is not name spaced correctly, we could have `dns_sd_handle_ptr_query()` or similar instead.